TreeView: Add support for leading and trailing visuals (#2383)
primer/react: docs/content/TreeView.mdx

> ### File tree navigation without directory links

Reviewer: Could we get an example (and maybe a story?) showing how to use the `isExpanded` render prop to set different icons for expanded and collapsed states.

Author: Great idea 👍 Let me know if the docs I added make sense

```jsx live drafts
<Box sx={{maxWidth: 320}}>
  <nav aria-label="File navigation">
    <TreeView aria-label="File navigation">
      <TreeView.Item>
        <TreeView.LeadingVisual>
          <TreeView.DirectoryIcon />
        </TreeView.LeadingVisual>
        src
        <TreeView.SubTree>
          <TreeView.LinkItem href="#">
            <TreeView.LeadingVisual>
              <FileIcon />
            </TreeView.LeadingVisual>
            Avatar.tsx
            <TreeView.TrailingVisual>

>               <StyledOcticon icon={DiffAddedIcon} color="success.fg" />

Reviewer: Since these trailing visuals are used to visually communicate diff status, I'm pretty sure they should have an `aria-label`.

cc @ericwbailey

Author: That makes sense. How should the `aria-label` for the trailing visual impact the `aria-label` of the treeitem?

Author: Currently, the aria-label would be `Avatar.tsx added`. That seems correct but I'd love confirmation from @ericwbailey.

Author: I updated the docs and stories to include `aria-label` on these trailing visuals 👍

Reviewer 2: Sorry to get to this after merging. I try to avoid using `aria-label` on non-interactive elements, given their mixed support. In the API guidance I am [recommending content that uses both `aria-hidden="true"` and a `sr-only` class applied to it](https://github.com/github/primer/pull/1321/files#diff-6d582e9797a0fa523d2bb3ed0ba3ed1739200d6d3b37b9e649fd348c494a9a2bR842-R847). 

In addition to the mixed support concern, we're also going to be using the text string to construct `aria-labeledby` announcements for the relevant TreeView node.

render(<ControlledTreeView />)
```

### With stateful visuals

> To render stateful visuals, pass a render function to `TreeView.LeadingVisual` and `TreeView.TrailingVisual`. The function will be called with the `expanded` state of the item.

Author: ```suggestion
To render stateful visuals, you can pass a render function to `TreeView.LeadingVisual` and `TreeView.TrailingVisual`. The function will be called with the `expanded` state of the item.
```

```jsx live drafts
<Box sx={{maxWidth: 400}}>
  <nav aria-label="File navigation">
    <TreeView aria-label="File navigation">
      <TreeView.Item>
        <TreeView.LeadingVisual>
          {({isExpanded}) => (isExpanded ? <FileDirectoryOpenFillIcon /> : <FileDirectoryFillIcon />)}
        </TreeView.LeadingVisual>
        src
        <TreeView.SubTree>
          <TreeView.LinkItem href="#">Avatar.tsx</TreeView.LinkItem>
          <TreeView.LinkItem href="#" current>
            Button.tsx
          </TreeView.LinkItem>
        </TreeView.SubTree>
      </TreeView.Item>
    </TreeView>
  </nav>
</Box>
```

> Since stateful directory icons are a common usecase for TreeView, we provide a `TreeView.DirectoryIcon` component for convenience. The above example can be rewritten as:

Author: ```suggestion
Since stateful directory icons are a common use case for TreeView, we provide a `TreeView.DirectoryIcon` component for convenience. The previous example can be rewritten as:
```

    hasFigmaComponent: false
  }}
/>

> ````

Author: ```suggestion

```
